### PR TITLE
backend electro shop

### PR DESCRIPTION
**Pull request by MykolaRudenko7**

### CODE REVIEW

File: src/controllers/cart-controller.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import ApiError from '../exceptions/api-error.js'
```
Comment (yurii-kozlov) [inline src/controllers/cart-controller.js]: it would be good to set absolute imports

File: src/controllers/cart-controller.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+class CartController {
+  async shippingHandler(req, res, next) {
```
Comment (yurii-kozlov) [inline src/controllers/cart-controller.js]:
```suggestion
  async handleShipping(req, res, next) {
```
Also indicate please in the name of the method what is shipped

File: src/controllers/products-controller.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import ApiError from '../exceptions/api-error.js'
```
Comment (yurii-kozlov) [inline src/controllers/products-controller.js]: please set up absolute imports

File: src/data/config.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+export const tokenConfig = {
+  accessTokenKey: process.env.ACCESS_TOKEN_KEY,
+  refreshTokenKey: process.env.REFRESH_TOKEN_KEY,
+  accessTokenExpires: process.env.ACCESS_TOKEN_EXPIRES_IN,
+  refreshTokenExpires: process.env.REFRESH_TOKEN_EXPIRES_IN,
+}
+
+export const appAddresses = {
```
Comment (yurii-kozlov) [inline src/data/config.js]:
```suggestion
export const appEndpoints = {
```

please move it into a separate file as it is not a config

File: src/controllers/user-controller.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+import TokenService from '../service/token-service.js'
+import UserService from '../service/user-service.js'
+import ApiError from '../exceptions/api-error.js'
+
+const { client } = appAddresses
+
+class UserController {
+  async signUp(req, res, next) {
+    try {
+      const validateErrors = validationResult(req)
```
Comment (yurii-kozlov) [inline src/controllers/user-controller.js]:
```suggestion
      const validationErrors = validationResult(req)
```

File: src/controllers/user-controller.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      return TokenService.handleSetResponse(res, userData, 'Registration was successful')
+    } catch (error) {
+      next(error)
+    }
+  }
+
+  async activateUser(req, res, next) {
+    try {
+      const activationLink = req.params.link
+      await UserService.activate(activationLink)
```
Comment (yurii-kozlov) [inline src/controllers/user-controller.js:35]: does this method activate the account? Then it should be renamed to smth like 'confirm account' or 'activate account'

File: src/middlewares/error-middleware.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
+import ApiError from '../exceptions/api-error.js'
+
+export default function (err, req, res, next) {
+  console.log(err)
```
Comment (yurii-kozlov) [inline src/middlewares/error-middleware.js]: pls remove logging to console here

File: src/routes/index.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
+import { Router } from 'express'
+import { body } from 'express-validator'
+import ProductsController from '../controllers/products-controller.js'
+import UserController from '../controllers/user-controller.js'
+import CartController from '../controllers/cart-controller.js'
+import authMiddleware from '../middlewares/auth-middleware.js'
+
+const router = new Router()
+
+router.post(
```
Comment (yurii-kozlov) [inline src/routes/index.js]: pls move validation into a separate file 'userValidation.js'

File: src/data/config.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  smprtHost: process.env.SMPRT_HOST,
+  smprtPort: process.env.SMPRT_PORT,
+  mailAddress: process.env.MAIL_ADDRESS,
+  mailPassword: process.env.MAIL_PASSWORD,
+}
+
+export const tokenConfig = {
+  accessTokenKey: process.env.ACCESS_TOKEN_KEY,
+  refreshTokenKey: process.env.REFRESH_TOKEN_KEY,
+  accessTokenExpires: process.env.ACCESS_TOKEN_EXPIRES_IN,
```
Comment (yurii-kozlov) [inline src/data/config.js]:
```suggestion
  accessTokenExpirationTime: process.env.ACCESS_TOKEN_EXPIRES_IN,
```

File: src/data/config.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  refreshTokenExpires: process.env.REFRESH_TOKEN_EXPIRES_IN,
```
Comment (yurii-kozlov) [inline src/data/config.js]:
```suggestion
  refreshTokenExpirationTime: process.env.REFRESH_TOKEN_EXPIRES_IN,
```

File: src/service/token-service.js [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+      .cookie('accessToken', accessToken, {
+        httpOnly: true,
+        secure: true,
+        maxAge: 15 * 60 * 1000,
+      })
+
+    return (
+      res
+        // я тепер передаю тут і токени
+        .json({ success: true, message: message, user, accessToken, refreshToken })
```
Comment (yurii-kozlov) [inline src/service/token-service.js]: why do you need to pass tokens in the body? It seem that you pass them to cookies already

File: src/service/token-service.js [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+    }
+  }
+
+  async findRefreshToken(refreshToken) {
+    const isRefreshToken = await UserToken.findOne({ refreshToken })
+
+    return isRefreshToken
+  }
+
+  async handleSetResponse(res, userData, message) {
```
Comment (yurii-kozlov) [inline src/service/token-service.js]:
```suggestion
  async sendAuthResponseWithTokens(res, userData, message) {
```

File: src/service/user-service.js [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+
+    if (!user) {
+      throw ApiError.BadRequest('Incorrect activation link')
+    }
+
+    user.isActivated = true
+    await user.save()
+  }
+
+  async signInUser(email, password) {
```
Comment (yurii-kozlov) [inline src/service/user-service.js:60]: It seems that this method violates the SRP, that's why you need to create different functions for different responsibilities. Please consider creating separate methods:

- getUserByEmail
- handleInvalidUser
- handleIncorrectPassword
- resetLockAndLoginAttempts

**Review by yurii-kozlov — CHANGES_REQUESTED** [review]

File: src/controllers/cart-controller.js [hunk 14]
```diff
@@ -0,0 +1,17 @@
+import ApiError from '#exceptions/api-error.js'
+
+class CartController {
+  async handleShipping(req, res, next) {
```
Comment (yurii-kozlov) [inline src/controllers/cart-controller.js]: maybe 'handleProductsShipping' would be a better naming

File: src/service/user-service.js [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+import MailService from './mail-service.js'
+import { salt } from '#data/config.js'
+import { appEndpoints } from '#data/appEndpoints.js'
+import User from '#models/User.js'
+
+const { backend } = appEndpoints
+
+class UserService {
+  async signUpUser(name, email, password, mobileNumber) {
+    const userExistsByEmail = await User.findOne({ email })
```
Comment (yurii-kozlov) [inline src/service/user-service.js]:
```suggestion
    const doesUserExistByEmail = await User.findOne({ email })
```

File: src/service/user-service.js [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+    const userExistsByPhone = await User.findOne({ mobileNumber })
```
Comment (yurii-kozlov) [inline src/service/user-service.js]:
```suggestion
    const doesUserExistByPhone = await User.findOne({ mobileNumber })
```

File: src/utils/userValidation.js [hunk 17]
```diff
@@ -0,0 +1,18 @@
+import { body, validationResult } from 'express-validator'
+
+export const userValidationRules = () => {
```
Comment (yurii-kozlov) [inline src/utils/userValidation.js]: Do you need to create a function for this?

**Review by yurii-kozlov — APPROVED** [review]

**Review by RuchkoVlad — APPROVED** [review]

**Review by RuchkoVlad — APPROVED** [review]

**Pull request merged**